src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -3247,6 +3255,166 @@ closedir(DIR *dirp)
 	}
 }
 
+long
+telldir(DIR *dirp)
+{
+	int fd;
+
+	if (next_telldir == NULL) {
+		next_telldir = dlsym(RTLD_NEXT, "telldir");
+		D_ASSERT(next_telldir != NULL);
+	}
+	if (!hook_enabled)
+		return next_telldir(dirp);
+
+	fd = dirfd(dirp);
+	if (fd < FD_DIR_BASE)
+		return next_telldir(dirp);
+
+	return dir_list[fd - FD_DIR_BASE]->offset;
+}
+
+void
+rewinddir(DIR *dirp)
+{
+	int fd, idx;
+
+	if (next_rewinddir == NULL) {
+		next_rewinddir = dlsym(RTLD_NEXT, "rewinddir");
+		D_ASSERT(next_rewinddir != NULL);
+	}
+	if (!hook_enabled)
+		return next_rewinddir(dirp);
+
+	fd = dirfd(dirp);
+	if (fd < FD_DIR_BASE)
+		return next_rewinddir(dirp);
+
+	idx                     = fd - FD_DIR_BASE;
+	dir_list[idx]->offset   = 0;
+	dir_list[idx]->num_ents = 0;
+	memset(&dir_list[idx]->anchor, 0, sizeof(daos_anchor_t));
+
+	return;
+}
+
+/* Offset of the first entry, allow two entries for . and .. */
+#define OFFSET_BASE 2
+
+void
+seekdir(DIR *dirp, long loc)
+{
+	int      fd, idx, rc;
+	long     num_entry;
+	uint32_t num_to_read;
+
+	if (next_seekdir == NULL) {
+		next_seekdir = dlsym(RTLD_NEXT, "seekdir");
+		D_ASSERT(next_seekdir != NULL);
+	}
+	if (!hook_enabled)
+		return next_seekdir(dirp, loc);
+
+	fd = dirfd(dirp);
+	if (fd < FD_DIR_BASE)
+		return next_seekdir(dirp, loc);
+
+	idx = fd - FD_DIR_BASE;
+
+	/* need to compare loc with current offset & the number of cached entries */
```

Comment:
it's not clear, but i would like to see a test that covers each of those cases.
ideally one of the dfuse tests should have had that, but im not sure it exists.
can be added in another PR.

Reply:
I think the new code in dfuse_test.c does cover this.
https://github.com/daos-stack/daos/blob/lei/DAOS-14344/src/tests/suite/dfuse_test.c#L534-L550

Reply:
READ_DIR_BATCH_SIZE is defined as 96. There are 100 files under the tested directory. We check all possible valid offset.